### ISSUE TITLE
fast path for device_get of fully replicated SDA

### ISSUE THREAD

**Issue by jekbradbury** [google/jax#9044]
Some `ShardedDeviceArray`s (such as, frequently, globally accumulated metrics like loss and accuracy) are fully replicated, and calling `device_get` should only need to pull one replica back to the host. (I thought this would be fairly straightforward, but now that SDA is in C++ I really don't have much familiarity with the code, so I'm not working on this right now. I don't even know for sure that this isn't done already!)

**Comment by zhangqiaorjc** [comment]
@jekbradbury were you observing multiple redundant d2h copies from profiler?

`jax.device_get(ShardedDeviceArray)` ultimately calls `_sda_value(self)` which calls `copy_to_host_async()`, for C++ ShardedDeviceArray at least, that's  overridden dynamically by `_sda_copy_to_host_async(self)`

```
def _sda_copy_to_host_async(self):
  for buffer_index in self.one_replica_buffer_indices:
    self.device_buffers[buffer_index].copy_to_host_async()
```

https://github.com/google/jax/blob/main/jax/interpreters/pxla.py#L625-L627

For a fully replicated SDA, `self.sharding_spec= ShardingSpec((NoSharding(), NoSharding()), (Replicated(replicas=4), Replicated(replicas=2)))`  and `self.one_replica_buffer_indices=[0]`.

My understanding is that only one device_buffer is really being pulled from device if the SDA is fully replicated.

cc @hawkinsp

**Comment by jekbradbury** [comment]
Is this true in Pathways?

**Comment by zhangqiaorjc** [comment]
probably not for Pathways...it has a different ShardedDeviceArray impl that might be kept 100% in sync with JAX' SDA.

**Comment by zhangqiaorjc** [comment]
I searched for `one_replica_buffer_indices` in the repo, and saw one callsite, and i don't honestly know where`tfcpu.py` is used..

**Comment by zhangqiaorjc** [comment]
Let's close this one? And file it for Pathways?
